volatility3/framework/plugins/linux/capabilities.py

```diff
@@ -0,0 +1,228 @@
+        The format is a comma-separated list of capabilitites. In order to
+        summarize the output and if all the capabilities are enabled, instead of
+        the individual capabilities, the special name "all" will be shown.
+
+        Args:
+            cap: Kernel capability object. Usually a 'kernel_cap_struct' struct
+
+        Returns:
+            str: A string with a comma separated list of decoded capabilities
+        """
+        if isinstance(cap, renderers.NotAvailableValue):
+            return cap
+
+        cap_value = cap.get_capabilities()
+        if cap_value == 0:
+            return "-"
+
+        CAP_FULL = 0xFFFFFFFF
+        if cap_value == CAP_FULL:
+            return "all"
+
+        return ", ".join(cap.enumerate_capabilities())
+
+    @classmethod
+    def get_task_capabilities(cls, task: interfaces.objects.ObjectInterface) -> Dict:
+        """Returns a dict with the task basic information along with its capabilities
+
+        Args:
+            task: A task object from where to get the fields.
+
+        Returns:
+            dict: A dict with the task basic information along with its capabilities
+        """
+        task_cred = task.real_cred
+        fields = {
+            "common": [
+                utility.array_to_string(task.comm),
+                int(task.pid),
+                int(task.tgid),
+                int(task.parent.pid),
+                int(task.cred.euid),
+            ],
+            "capabilities": [
+                task_cred.cap_inheritable,
+                task_cred.cap_permitted,
+                task_cred.cap_effective,
+                task_cred.cap_bset,
+            ],
+        }
+
+        # Ambient capabilities were added in kernels 4.3.6
+        if task_cred.has_member("cap_ambient"):
+            fields["capabilities"].append(task_cred.cap_ambient)
+        else:
+            fields["capabilities"].append(renderers.NotAvailableValue())
+
+        return fields
+
+    def get_tasks_capabilities(
+        self, tasks: List[interfaces.objects.ObjectInterface]
+    ) -> Iterable[Dict]:
+        """Yields a dict for each task containing the task's basic information along with its capabilities
+
+        Args:
+            tasks: An iterable with the tasks to process.
+
+        Yields:
+            Iterable[Dict]: A dict for each task containing the task's basic information along with its capabilities
+        """
+        for task in tasks:
+            if task.is_kernel_thread:
+                continue
+
+            yield self.get_task_capabilities(task)
+
+    def _generator(
+        self, tasks: Iterable[interfaces.objects.ObjectInterface]
+    ) -> Iterable[Tuple[int, Tuple]]:
+        for fields in self.get_tasks_capabilities(tasks):
+            selected_fields = fields["common"]
+            cap_inh, cap_prm, cap_eff, cap_bnd, cap_amb = fields["capabilities"]
+
+            if self.config.get("inheritable"):
```

Comment:
Ewwwwww...  5:P  So it looks like this is completely changing the fields that are output from this function.  That somewhat goes against how we've done things previously, which is to return all data.  It's not as good for human readability, but it makes everything much cleaner and easier when used as a library.  Since these are binary values, it should only be a T or F (and if it says True/False, I'd sooner introduce a separate pull request that changes that than import all this branching).

I'm ok when it's a new capability in linux might change the capabilties returned, but a system that calls this plugin to populate a database is going to have a hell of time if ithe user's allowed to choose their options, and it's not really terrible if we spit everything back, as long as a user has a fairly easy way to limit the results.  I'd even sooner accept a PR to allow the CLI to specific which columns are printed.  Doing it within the plugin takes the ability to make display/formatting decisions away from the interface (which may need to make decisions like that)...

Reply:
Hey @ikelos thanks for the review. Let's start with this comment because depending on what we decide here many of the other comments may no longer apply. 

I'm not sure what you mean with the T/F, there are 5 capabilities categories and each of them can have 32 capabilities so I don't think you mean to have 160 fixed columns with true/false.

If you don't like this plugin changes the columns based on its command line parameters, I see two options here  
1) Following your second paragraph, we can show each of the 5 categories with its capabilities in text format 5 x (chown, dac_override, dac_read_search, etc.). This will be quite wide but as far as I understand there will be in the future a way to generically select which columns to show for any plugin in vol3, right? 

2) The most condensed way to show everything is actually the default output of this plugin which is the first plugin output showed above in the main PR description.
```shell
  |            Name |  Tid |  Pid | PPid | EUID | cap_inheritable | cap_permitted | cap_effective | cap_bounding | cap_ambient
* |         systemd |    1 |    1 |    0 |    0 |             0x0 |    0xffffffff |    0xffffffff |   0xffffffff |         0x0
* | systemd-journal |  456 |  456 |    1 |    0 |             0x0 |    0x402800cf |    0x402800cf |   0x402800cf |         0x0
...
```
But in order to an human being read that, they will need to used the function helpers I also added to this PR, or to use externals tools like the **capsh** decoder  i.e: capsh --decode=0x402800cf .. but I see this very impractical and that's why I included options so that the user can have the desired output.

Last but not least, I don't see why a system that calls this plugin will have a hell of time. Actually, calling this plugin as a library is straight forward. They can call  get_task_capabilities(task) for an individual task or get_tasks_capabilities(tasks) (yes, I will make this a class_method ;) ) to get the capabilities of all the tasks. 
The returned format for each task is a dictionary as below. 
```python
{
            "common": [
                comm,
                pid,
                tgid,
                pid,
                euid,
            ],
            "capabilities": [
                cap_inheritable,
                cap_permitted,
                cap_effective,
                cap_bset,
                cap_ambient
            ],
}
```
Then, for each cap_* it can call cap_foo.enumerate_capabilities() and it gets the textual representation, as it's done in the generator. The plugin options are only taken into account when the generator is executed. A system using this plugin as a library don't need to use the generator.

Reply:
Hiya, I didn't realize there were so many possibilities!  5:S  In that instance then it does change things a little. 160 columns would be too many, I agree.  The textual output is probably for the best here, but I would like all the columns output every time (if there's no content, they should return a `BaseAbsentValue`).  This saves a framework from having to parse the TreeGrid and special case specific columns being present or not.

The output will be wide on the CLI, but another UI could display that data differently (a web app may choose to allow it to be expanded, etc).  It should still be easy enough for a human to read if they're names from an enumeration

Choosing columns isn't currently a planned feature for the CLI, but we absolutely could, the point is that's in the hands of the UI, not in the hands of the plugin.  Whilst the plugin does provide specific functions for doing things, a UI should not need to know about them to get the data from volatility.  The hope is that UIs need as little "special casing" for the outputs of plugins as possible (hence not requiring code if the number of columns change).  This is specifically generic UIs, not tools using volatility as library.  Variable column names, having to interpret data, or having to parse/post-process data all go against that idea.  I think commas is the least bad thing we can have in this instance where there is so much dense information that needs displaying.

I hope that all makes sense and you're happy with the reasoning behind it?  I may entirely still have missed something, or not considered the different between cases where everything is turned on and nothing is turned on.  If there's still something you want me to think over or a case I've missed, just let me know, I'm happy to keep talking this through until we get it right.  5:)

( I agree the T/F thing was not the right choice for this data, though, thanks for catching me on it!)  5:D

Reply:
Ok, I had this ready waiting you reply :) ... Have a look now. It looks much cleaner. I used [dataclasses](https://docs.python.org/3.7/library/dataclasses.html) introduced in Python 3.7 (all good with vol3 minimum requirements) which are really cool, not like namedtuples.